https://mp.weixin.qq.com/s?__biz=MzA3MzI4MjgzMw==&mid=2650780679&idx=5&sn=713353a7129b040668273e7473169e3b&chksm=871a7079b06df96f742a241c6cad948cc487fb9ffff05c3010c5798371ff878eeac50246ffa5&mpshare=1&scene=1&srcid=&sharer_sharetime=1582092115443&sharer_shareid=bd9bb1dccc48fe5d66218528b0778f2f&key=9c0a8c4c60b2594eae8f08ba1f3c8f17e913981a7f65f08745f5c9b10fac02ed2c6ddba36db19e627d30079832e064f4f765b80df126aca0a3eaae10f3369bf174659044208fd2625ee6dd08bcc3b69a&ascene=1&uin=MTc2MDIyOTE2OA%3D%3D&devicetype=Windows+10&version=62070158&lang=zh_CN&exportkey=AapujGnrfiskBRzGfW17JEo%3D&pass_ticket=A61l49hrOE6v2dP4hcIQ5kWWskJHer8MWeNApsLDBI4bS0VRUyfkswFhFwSKW5cV

下图是乳腺癌 DataFrame 的一部分。每行是一个患者的观察结果。最后一列名为「cancer」是我们要预测的目标变量。0 表示「无癌症」，1 表示「癌症」。

![jupyter](.\breast.png)

## 导入数据

首先，我们加载数据并创建一个 DataFrame。这是 Scikit-learn 预先清理的「toy」数据集，因此我们可以继续快速建模。但是，作为最佳实践，我们应该执行以下操作：

- 使用 df.head（）查看新的 DataFrame，以确保它符合预期。

- 使用 df.info（）可以了解每一列中的数据类型和数据量。可能需要根据需要转换数据类型。

- 使用 df.isna（）确保没有 NaN 值。可能需要根据需要处理缺失值或删除行。

- 使用 df.describe（）可以了解每列的最小值、最大值、均值、中位数、标准差和四分位数范围。


名为「cancer」的列是我们要使用模型预测的目标变量。「0」表示「无癌症」，「1」表示「癌症」。

In [7]:
import pandas as pd
# 从sklearn导入数据包
from sklearn.datasets import load_breast_cancer
# 要导入的数据列
columns = ['mean radius', 'mean texture', 'mean perimeter', 'mean area'
           , 'mean smoothness', 'mean compactness', 'mean concavity'
           , 'mean concave points', 'mean symmetry', 'mean fractal dimension'
           , 'radius error', 'texture error', 'perimeter error', 'area error'
           , 'smoothness error', 'compactness error', 'concavity error'
           , 'concave points error', 'symmetry error', 'fractal dimension error'
           , 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 
           'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points'
           , 'worst symmetry', 'worst fractal dimension']
# 导入数据包
dataset = load_breast_cancer()
# 导入数据
data = pd.DataFrame(dataset['data'], columns=columns)
# 目标列
data['cancer'] = dataset['target']
display(data.head())
display(data.info())
display(data.isna().sum())
display(data.describe())

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  cancer  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error                 569 non-null float64
smoothness error           569 non-null float64
compactness error          569 non-null float64
concavity error            569 non-null float64
concave points error       569 non-null float64
symmetry error             569 

None

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
cancer                     0
dtype: int64

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      cancer  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

## 训练集/测试集分割

现在，我们使用 Scikit-learn 的「train_test_split」函数拆分数据。我们想让模型有尽可能多的数据进行训练。但是，我们也要确保有足够的数据来测试模型。通常数据集中行数越多，我们可以提供给训练集的数据越多。

例如，如果我们有数百万行，那么我们可以将其中的 90％用作训练，10％用作测试。但是，我们的数据集只有 569 行，数据量并不大。因此，为了匹配这种小型数据集，我们会将数据分为 50％的训练和 50％的测试。我们设置 stratify = y 以确保训练集和测试集与原始数据集的 0 和 1 的比例一致。

In [16]:
from sklearn.model_selection import train_test_split
# 取删除'cancer'列后的数据
X = data.drop('cancer', axis=1)
y = data['cancer']
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.50, random_state = 2020, stratify=y)
x_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
250       20.940         23.56          138.90     1364.0          0.10070   
409       12.270         17.92           78.41      466.1          0.08685   
524        9.847         15.68           63.00      293.2          0.09492   
203       13.810         23.75           91.56      597.8          0.13230   
435       13.980         19.62           91.12      599.5          0.10600   
..           ...           ...             ...        ...              ...   
206        9.876         17.27           62.92      295.4          0.10890   
308       13.500         12.71           85.69      566.2          0.07376   
187       11.710         17.19           74.68      420.3          0.09774   
359        9.436         18.32           59.82      278.6          0.10090   
421       14.690         13.98           98.22      656.1          0.10310   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
250           0.16060        0.271200             0.131000         0.2205   
409           0.06526        0.032110             0.026530         0.1966   
524           0.08419        0.023300             0.024160         0.1387   
203           0.17680        0.155800             0.091760         0.2251   
435           0.11330        0.112600             0.064630         0.1669   
..                ...             ...                  ...            ...   
206           0.07232        0.017560             0.019520         0.1934   
308           0.03614        0.002758             0.004419         0.1365   
187           0.06141        0.038090             0.032390         0.1516   
359           0.05956        0.027100             0.014060         0.1506   
421           0.18360        0.145000             0.063000         0.2086   

     mean fractal dimension  ...  worst radius  worst texture  \
250                 0.05898  ...         25.58          27.00   
409                 0.05597  ...         14.10          28.88   
524                 0.06891  ...         11.24          22.99   
203                 0.07421  ...         19.20          41.85   
435                 0.06544  ...         17.04          30.80   
..                      ...  ...           ...            ...   
206                 0.06285  ...         10.42          23.22   
308                 0.05335  ...         14.97          16.94   
187                 0.06095  ...         13.01          21.39   
359                 0.06959  ...         12.02          25.02   
421                 0.07406  ...         16.46          18.34   

     worst perimeter  worst area  worst smoothness  worst compactness  \
250           165.30      2010.0           0.12110            0.31720   
409            89.00       610.2           0.12400            0.17950   
524            74.32       376.5           0.14190            0.22430   
203           128.50      1153.0           0.22260            0.52090   
435           113.90       869.3           0.16130            0.35680   
..               ...         ...               ...                ...   
206            67.08       331.6           0.14150            0.12470   
308            95.48       698.7           0.09023            0.05836   
187            84.42       521.5           0.13230            0.10400   
359            75.79       439.6           0.13330            0.10490   
421           114.10       809.2           0.13120            0.36350   

     worst concavity  worst concave points  worst symmetry  \
250          0.69910               0.21050          0.3126   
409          0.13770               0.09532          0.3455   
524          0.08434               0.06528          0.2502   
203          0.46460               0.20130          0.4432   
435          0.40690               0.18270          0.3179   
..               ...                   ...             ...   
206          0.06213               0.05588         

## 规范化数据

在建模之前，我们需要先将数据「居中」和「标准化」，对不同的变量要在相同尺度进行测量。我们进行缩放以便决定预测变量的特征可以彼此「公平竞争」。我们还将「y_train」从 Pandas「Series」对象转换为 NumPy 数组，以供模型稍后接收训练数据。

In [15]:
import numpy as np
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
x_train_scaled = ss.fit_transform(x_train)
x_test_scaled = ss.transform(x_test)
y_train = np.array(y_train)
pd.DataFrame(x_train_scaled, columns = x_train.columns)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0       1.859207      0.963565        1.848631   1.877184         0.209662   
1      -0.517899     -0.307403       -0.546357  -0.515278        -0.750872   
2      -1.182228     -0.812184       -1.156488  -0.975972        -0.191196   
3      -0.095668      1.006381       -0.025708  -0.164363         2.401205   
4      -0.049058      0.075690       -0.043129  -0.159833         0.577231   
..           ...           ...             ...        ...              ...   
279    -1.174277     -0.453880       -1.159655  -0.970110         0.778354   
280    -0.180663     -1.481471       -0.258119  -0.248561        -1.658697   
281    -0.671438     -0.471908       -0.694040  -0.637313         0.004378   
282    -1.294914     -0.217263       -1.282394  -1.014874         0.223533   
283     0.145607     -1.195278        0.237983  -0.009022         0.376109   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0            1.023032        2.170616             1.951691       1.484489   
1           -0.760717       -0.693578            -0.593879       0.575067   
2           -0.406549       -0.799117            -0.651628      -1.628091   
3            1.326123        0.788174             0.995549       1.659524   
4            0.138080        0.270657             0.334485      -0.555050   
..                ...             ...                  ...            ...   
279         -0.628629       -0.867880            -0.764689       0.453304   
280         -1.305533       -1.045202            -1.132648      -1.711804   
281         -0.832748       -0.621940            -0.451092      -1.137232   
282         -0.867361       -0.753595            -0.897730      -1.175283   
283          1.453346        0.658795             0.294768       1.031681   

     mean fractal dimension  ...  worst radius  worst texture  \
0                 -0.585456  ...      1.883610       0.229568   
1                 -1.029340  ...     -0.440226       0.532662   
2                  0.878921  ...     -1.019161      -0.416924   
3                  1.660512  ...      0.592140       2.623684   
4                  0.367200  ...      0.154903       0.842204   
..                      ...  ...           ...            ...   
279               -0.014747  ...     -1.185149      -0.379843   
280               -1.415712  ...     -0.264117      -1.392304   
281               -0.294940  ...     -0.660869      -0.674876   
282                0.979200  ...     -0.861270      -0.089647   
283                1.638391  ...      0.037496      -1.166596   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0           1.696719    1.894926         -0.565515           0.424972   
1          -0.529988   -0.458521         -0.438955          -0.466276   
2          -0.958403   -0.851435          0.342222          -0.176313   
3           0.622763    0.454074          3.864069           1.743398   
4           0.196683   -0.022903          1.188862           0.681279   
..               ...         ...               ...                ...   
279        -1.169692   -0.926924          0.324766          -0.820963   
280        -0.340879   -0.309728         -1.912719          -1.250341   
281        -0.663649   -0.607650         -0.076733          -0.954941   
282        -0.915503   -0.745346         -0.033092          -0.949116   
283         0.202520   -0.123948         -0.124739           0.724644   

     worst concavity  worst concave points  worst symmetry  \
0           2.060807              1.417609        0.365792   
1          -0.618008             -0.312313        0.904378   
2          -0.872624             -0.763492       -0.655721   
3           0.941850              1.279431        2.503765   
4           0.666525              1.000072        0.452555   
..               ...                   ...             ...   
279        -0.978603             -0.904674        0

## 拟合「基线」随机森林模型

现在，我们创建一个「基线」随机森林模型。该模型使用 Scikit-learn 随机森林分类器文档中定义的所有预测特征和默认设置。首先，我们实例化模型并使用规范化的数据拟合模型。我们可以通过训练数据测量模型的准确性。

In [47]:
from sklearn.ensemble import RandomForestClassifier    # 导入随机森林分类器
from sklearn.metrics import recall_score
rfc = RandomForestClassifier()
rfc.fit(x_train_scaled, y_train)
display(rfc.score(x_train_scaled, y_train))

1.0

如果我们想知道哪些特征对随机森林模型预测乳腺癌最重要，我们可以通过调用「feature_importances _」方法来可视化和量化这些重要特征：

# 什么叫[基线模型]???

In [34]:
import numpy as np
from sklearn.model_selection import KFold,StratifiedKFold
 
# 8 x 4 
X = np.array([
    [1,2,3,4],
    [11,12,13,14],
    [21,22,23,24],
    [31,32,33,34],
    [41,42,43,44],
    [51,52,53,54],
    [61,62,63,64],
    [71,72,73,74]
])

# 正类4 负类4
y = np.array([1,1,0,0,1,1,0,0])

kf = KFold(n_splits = 4, random_state=0, shuffle = False)
skf = StratifiedKFold(n_splits = 4, random_state = 0, shuffle = False)

print('StraifiedKFold：')
for train_index_s, test_index_s in skf.split(X,y):     # y必须传入，因为要按类别分层
    print(f'train_index_s:{train_index_s}| test_index_s:{test_index_s}')
    

print('\nKFold：')
for train_index_s, test_index_s in kf.split(X,y):     # y传入不传入没多大关系，随机划分
    print(f'train_index_s:{train_index_s}| test_index_s:{test_index_s}')
    
# 如何取得训练集和测试集呢？
# 训练集：X[train_index_s]    y[train_index_s]
# 测试集：X[test_index_s]     y[test_index_s]

StraifiedKFold：
train_index_s:[1 3 4 5 6 7]| test_index_s:[0 2]
train_index_s:[0 2 4 5 6 7]| test_index_s:[1 3]
train_index_s:[0 1 2 3 5 7]| test_index_s:[4 6]
train_index_s:[0 1 2 3 4 6]| test_index_s:[5 7]

KFold：
train_index_s:[2 3 4 5 6 7]| test_index_s:[0 1]
train_index_s:[0 1 4 5 6 7]| test_index_s:[2 3]
train_index_s:[0 1 2 3 6 7]| test_index_s:[4 5]
train_index_s:[0 1 2 3 4 5]| test_index_s:[6 7]


In [10]:
from sklearn.model_selection import KFold
import numpy as np
X = np.arange(24).reshape(12,2)
kf = KFold(n_splits=5,shuffle=False)
for train_index , test_index in kf.split(X):
    print('train_index:%s , test_index: %s ' %(train_index,test_index))

train_index:[ 3  4  5  6  7  8  9 10 11] , test_index: [0 1 2] 
train_index:[ 0  1  2  6  7  8  9 10 11] , test_index: [3 4 5] 
train_index:[ 0  1  2  3  4  5  8  9 10 11] , test_index: [6 7] 
train_index:[ 0  1  2  3  4  5  6  7 10 11] , test_index: [8 9] 
train_index:[0 1 2 3 4 5 6 7 8 9] , test_index: [10 11] 


In [11]:
X

array([[ 0,  1],
       [ 2,  3],
       [ 4,  5],
       [ 6,  7],
       [ 8,  9],
       [10, 11],
       [12, 13],
       [14, 15],
       [16, 17],
       [18, 19],
       [20, 21],
       [22, 23]])

In [21]:
from sklearn.model_selection import train_test_split,cross_val_score,cross_validate # 交叉验证所需的函数
from sklearn.model_selection import KFold,LeaveOneOut,LeavePOut,ShuffleSplit # 交叉验证所需的子集划分方法
from sklearn.model_selection import StratifiedKFold,StratifiedShuffleSplit # 分层分割
from sklearn.model_selection import GroupKFold,LeaveOneGroupOut,LeavePGroupsOut,GroupShuffleSplit # 分组分割
from sklearn.model_selection import TimeSeriesSplit # 时间序列分割
from sklearn import datasets  # 自带数据集
from sklearn import svm  # SVM算法
from sklearn import preprocessing  # 预处理模块
from sklearn.metrics import recall_score  # 模型度量

iris = datasets.load_iris()  # 加载数据集
print('样本集大小：',iris.data.shape,iris.target.shape)

# ===================================数据集划分,训练模型==========================
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.4, random_state=0)  # 交叉验证划分训练集和测试集.test_size为测试集所占的比例
print('训练集大小：',X_train.shape,y_train.shape)  # 训练集样本大小
print('测试集大小：',X_test.shape,y_test.shape)  # 测试集样本大小
clf = svm.SVC(kernel='linear', C=1).fit(X_train, y_train) # 使用训练集训练模型
print('准确率：',clf.score(X_test, y_test))  # 计算测试集的度量值（准确率）


#  如果涉及到归一化，则在测试集上也要使用训练集模型提取的归一化函数。
scaler = preprocessing.StandardScaler().fit(X_train)  # 通过训练集获得归一化函数模型。（也就是先减几，再除以几的函数）。在训练集和测试集上都使用这个归一化函数
X_train_transformed = scaler.transform(X_train)
clf = svm.SVC(kernel='linear', C=1).fit(X_train_transformed, y_train) # 使用训练集训练模型
X_test_transformed = scaler.transform(X_test)
print(clf.score(X_test_transformed, y_test))  # 计算测试集的度量值（准确度）

# ===================================直接调用交叉验证评估模型==========================
clf = svm.SVC(kernel='linear', C=1)
scores = cross_val_score(clf, iris.data, iris.target, cv=5)  #cv为迭代次数。
print(scores)  # 打印输出每次迭代的度量值（准确度）
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))  # 获取置信区间。（也就是均值和方差）

# ===================================多种度量结果======================================
scoring = ['precision_macro', 'recall_macro'] # precision_macro为精度，recall_macro为召回率
scores = cross_validate(clf, iris.data, iris.target, scoring=scoring,cv=5, return_train_score=True)
sorted(scores.keys())
print('测试结果：',scores)  # scores类型为字典。包含训练得分，拟合次数， score-times （得分次数）


# ==================================K折交叉验证、留一交叉验证、留p交叉验证、随机排列交叉验证==========================================
# k折划分子集
kf = KFold(n_splits=4)
for train, test in kf.split(iris.data):
    print("k折划分：%s %s" % (train.shape, test.shape))
    break

# 留一划分子集
loo = LeaveOneOut()
for train, test in loo.split(iris.data):
    print("留一划分：%s %s" % (train.shape, test.shape))
    break

# 留p划分子集
lpo = LeavePOut(p=2)
for train, test in loo.split(iris.data):
    print("留p划分：%s %s" % (train.shape, test.shape))
    break

# 随机排列划分子集
ss = ShuffleSplit(n_splits=3, test_size=0.25,random_state=0)
for train_index, test_index in ss.split(iris.data):
    print("随机排列划分：%s %s" % (train.shape, test.shape))
    break

# ==================================分层K折交叉验证、分层随机交叉验证==========================================
skf = StratifiedKFold(n_splits=3)  #各个类别的比例大致和完整数据集中相同
for train, test in skf.split(iris.data, iris.target):
    print("分层K折划分：%s %s" % (train.shape, test.shape))
    break

skf = StratifiedShuffleSplit(n_splits=3)  # 划分中每个类的比例和完整数据集中的相同
for train, test in skf.split(iris.data, iris.target):
    print("分层随机划分：%s %s" % (train.shape, test.shape))
    break


# ==================================组 k-fold交叉验证、留一组交叉验证、留 P 组交叉验证、Group Shuffle Split==========================================
X = [0.1, 0.2, 2.2, 2.4, 2.3, 4.55, 5.8, 8.8, 9, 10]
y = ["a", "b", "b", "b", "c", "c", "c", "d", "d", "d"]
groups = [1, 1, 1, 2, 2, 2, 3, 3, 3, 3]

# k折分组
gkf = GroupKFold(n_splits=3)  # 训练集和测试集属于不同的组
for train, test in gkf.split(X, y, groups=groups):
    print("组 k-fold分割：%s %s" % (train, test))

# 留一分组
logo = LeaveOneGroupOut()
for train, test in logo.split(X, y, groups=groups):
    print("留一组分割：%s %s" % (train, test))

# 留p分组
lpgo = LeavePGroupsOut(n_groups=2)
for train, test in lpgo.split(X, y, groups=groups):
    print("留 P 组分割：%s %s" % (train, test))

# 随机分组
gss = GroupShuffleSplit(n_splits=4, test_size=0.5, random_state=0)
for train, test in gss.split(X, y, groups=groups):
    print("随机分割：%s %s" % (train, test))


# ==================================时间序列分割==========================================
tscv = TimeSeriesSplit(n_splits=3)
TimeSeriesSplit(max_train_size=None, n_splits=3)
for train, test in tscv.split(iris.data):
    print("时间序列分割：%s %s" % (train, test))

样本集大小： (150, 4) (150,)
训练集大小： (90, 4) (90,)
测试集大小： (60, 4) (60,)
准确率： 0.9666666666666667
0.9333333333333333
[0.96666667 1.         0.96666667 0.96666667 1.        ]
Accuracy: 0.98 (+/- 0.03)
测试结果： {'fit_time': array([0.        , 0.        , 0.00099754, 0.        , 0.00099707]), 'score_time': array([0.0009973 , 0.0009973 , 0.00099754, 0.00099897, 0.        ]), 'test_precision_macro': array([0.96969697, 1.        , 0.96969697, 0.96969697, 1.        ]), 'train_precision_macro': array([0.97674419, 0.97674419, 0.99186992, 0.98412698, 0.98333333]), 'test_recall_macro': array([0.96666667, 1.        , 0.96666667, 0.96666667, 1.        ]), 'train_recall_macro': array([0.975     , 0.975     , 0.99166667, 0.98333333, 0.98333333])}
k折划分：(112,) (38,)
留一划分：(149,) (1,)
留p划分：(149,) (1,)
随机排列划分：(149,) (1,)
分层K折划分：(99,) (51,)
分层随机划分：(135,) (15,)
组 k-fold分割：[0 1 2 3 4 5] [6 7 8 9]
组 k-fold分割：[0 1 2 6 7 8 9] [3 4 5]
组 k-fold分割：[3 4 5 6 7 8 9] [0 1 2]
留一组分割：[3 4 5 6 7 8 9] [0 1 2]
留一组分割：[0 1 2 6 7 8 9] [3 